src/app/signup/page.tsx

```diff
@@ -1,14 +1,23 @@
 "use client";
 
-import Image from "next/image";
 import { GradButton } from "@/components/ui/grad-button";
 import SignUpImage from "@/components/ui/signup-image";
+import { SignUpFieldErrors, signupUser } from "@/lib/User/actions";
+import { EMAIL_ALREADY_EXISTS_MSG, signUpSchema } from "@/lib/User/validations";
 import { zodResolver } from "@hookform/resolvers/zod";
+import Image from "next/image";
+import { Fragment, useCallback, useRef, useState } from "react";
 import { useForm } from "react-hook-form";
-import { signupUser } from "@/lib/User/actions";
-import { signUpSchema } from "@/lib/User/validations";
 
 const SignUp = () => {
+  const formRef = useRef<HTMLFormElement>(null);
+  const [serverErrors, setServerErrors] = useState<
+    SignUpFieldErrors | undefined
+  >();
+  const [emailBeforeSubmission, setEmailBeforeSubmission] = useState<
```

Comment:
Do we still need this with the form fixed? It should keep its original value even after the server validation

Reply:
Well, if we want the functionality of removing the "email exists" error at some point and notify the user if they typed it again, then yes. We need to keep an ephemeral state. This is why I needed the flexibility I can't get from useFormState.

Reply:
I'll use Ref instead. A commit is coming up.